### PR TITLE
ROX-23953: Prefer digest from registry header if avail

### PR DESCRIPTION
**Pull request by dcaravel**
## Description

Scanner V4 requires an image to be referenced by digest, if a tag is passed to the Scanner V4 client an `unsupported digest algorithm` error will be returned.

Prior to this change ACS would set a `tag` to the `digest` field in the image metadata (which arguably should never happen).  This change increases the chances of that `digest` field containing an actual digest.

Builds on: 

- https://github.com/stackrox/stackrox/pull/10983
- https://github.com/stackrox/docker-registry-client/pull/20

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [x] Evaluated and added CHANGELOG entry if required
- [x] Determined and documented upgrade steps
- [x] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

Unit Tests and ...

On an OCP infra cluster - ensure Scanner V4 is enabled, and disable `ROX_ATTEMPT_MANIFEST_DIGEST` (this is done to simulate a registry that doesn't support the manifest `HEAD` call)

```sh
$ oc set env deploy/central ROX_ATTEMPT_MANIFEST_DIGEST=false ROX_SCANNER_V4=true
```

Before applying the fix, scan an image known to fail (this image has no manifest list and is referenced by tag, causing [this branch of code](https://github.com/stackrox/stackrox/blob/dbce8f374d55a066f5a793bc357539dce18baf5c/pkg/registries/docker/metadata_v2.go#L69) to be executed which sets the image's tag `7.1-24` into the `V2.Digest` field, ultimately breaking the Scanner V4 client.

```sh
$ rctl image scan --image=registry.access.redhat.com/rhel7.1:7.1-24 --force

ERROR:	Scanning image failed: could not scan image: "registry.access.redhat.com/rhel7.1:7.1-24": rpc error: code = Internal desc = image enrichment error: error scanning image: registry.access.redhat.com/rhel7.1:7.1-24 error: scanning "registry.access.redhat.com/rhel7.1:7.1-24" with scanner "Scanner V4": creating digest reference: unsupported digest algorithm: 7.1-24. Retrying after 3 seconds...
```

Then deploy the fix and retest:
```sh
$ oc set image deploy/central *=quay.io/rhacs-eng/main:4.4.x-671-g48e0fcf934

# wait for central reastart...

$ rctl image scan --image=registry.access.redhat.com/rhel7.1:7.1-24 --force

{
  "id": "sha256:9675a0cbb44ca8b2ef04f490ecd69242a60fdefdbca23881d66c3640c64ca37d",
  "name": {
    "registry": "registry.access.redhat.com",
    "remote": "rhel7.1",
    "tag": "7.1-24",
    "fullName": "registry.access.redhat.com/rhel7.1:7.1-24"
  },
...
```

### CODE REVIEW

File: pkg/registries/docker/metadata_v2.go [hunk 1]
```diff
@@ -102,8 +103,20 @@ func HandleOCIManifest(r *Registry, remote, ref string) (*storage.ImageMetadata,
...
+// Otherwise returns ref.
+func imageReference(ref string, digest godigest.Digest) string {
+	if digest != "" {
+		if _, err := godigest.Parse(ref); err != nil {
+			return string(digest)
+		}
+	}
+
+	return ref
+}
```
Comment (jvdm) [inline pkg/registries/docker/metadata_v2.go:122]: Maybe clearer:

```suggestion
// digestOrRef returns an image digest if found in the reference, then if found in the digest parameter.
// If no digest is found, return the reference as is.
func digestOrRef(ref string, digest godigest.Digest) string {
	_, err := godigest.Parse(ref)
	if err != nil && digest != "" {
		return string(digest)
	}	
	return ref
}
```

**Review by jvdm — APPROVED** [review]
LGTM, I have a suggestion, but it's a nitpick on style/clarity.

File: pkg/registries/docker/metadata_v2.go (hunk 1 above)
Comment (dcaravel) [reply pkg/registries/docker/metadata_v2.go:122]: I'll update the method name / comment, performing `godigest.Parse()` when digest == "" is wasteful however - rather leave that to be conditional.

Comment (dcaravel) [reply pkg/registries/docker/metadata_v2.go]: https://github.com/stackrox/stackrox/pull/11084/commits/f9ac114e1d571999cb0bea98e9d503ee2b4b47c7

Comment (dcaravel) [reply pkg/registries/docker/metadata_v2.go]: Changed up the logic to be clearer.

**Pull request merged**